// cf-autoconfig/lib/mongodb.js
/*
 * Cloud Foundry auto-configuration
 * Mongodb connection
 *
 * Auto-configured node modules:
 *
 * "mongodb" [https://github.com/mongodb/node-mongodb-native]
 * - connect()
 * - new Db()
 * - new Server()
 * - db.authenticate()
 *
 * "mongoose" [https://github.com/learnboost/mongoose]
 * Cause it can accept any drivers besides mongodb
 * - Connection.open() (most other functions depend on this)
 */

var sc = require(require("path").join(__dirname, "service"));
var util = require("util");

function Mongodb () {
  this.type = "mongodb";

  this.supportedModules = {
    "mongodb" : function (moduleData, props) {
      // connect()
      if ("connect" in moduleData) {
        var oldConnect = moduleData.connect;
        var oldConnectProto = moduleData.connect.prototype;

        moduleData.connect = function () {
          arguments[0] = props.url;
          return oldConnect.apply(this, arguments);
        };
        moduleData.connect.prototype = oldConnectProto;
      }

      // new Server()
      if ("Server" in moduleData) {
        var oldServer = moduleData.Server;
        var oldServerProto = moduleData.Server.prototype;

        moduleData.Server = function () {
          arguments[0] = props.host;
          arguments[1] = parseInt(props.port, 10);
          return oldServer.apply(this, arguments);
        };

        moduleData.Server.prototype = oldServerProto;
      }

      // new Db()
      if ("Db" in moduleData) {
        var oldDb = moduleData.Db;
        var oldDbProto = moduleData.Db.prototype;

        moduleData.Db = function () {
          var self = function () {
            arguments[0] = props.db;
            return oldDb.apply(this, arguments);
          };

          for (var prop in oldDb) {
            if (oldDb.hasOwnProperty(prop)) {
              self[prop] = oldDb[prop];
            }
          }

          return self;
        }();
        moduleData.Db.prototype = oldDbProto;

        // db.authenticate()
        if ("authenticate" in moduleData.Db.prototype) {
          var oldAuthenticate = moduleData.Db.prototype.authenticate;

          moduleData.Db.prototype.authenticate = function () {
            arguments[0] = props.username;
            arguments[1] = props.password;
            return oldAuthenticate.apply(this, arguments);
          };
        }

        // db.open() adds authenticate to callback
        if ("open" in moduleData.Db.prototype) {
          var oldOpen = moduleData.Db.prototype.open;

          moduleData.Db.prototype.open = function () {
            var callback = false;
            if (typeof arguments[0] === "function") callback = arguments[0];
            var self = this;
            oldOpen.call(this, function (err, db) {
              if (err === null) {
                db.authenticate.call(self, props.username, props.password, function(err, success) {
                  if(success) {
                    if (callback) callback.call(self, null, db);
                  } else {
                    if (callback)
                      callback.call(self, err ? err : new Error("Could not authenticate user"), db);
                  }
                });
              } else {
                if (callback) callback.call(self, err, db);
              }
            });
          };
        }
      }
    },

    "mongoose" : function (moduleData, props) {
      // Connection.open()
      if ("Connection" in moduleData && "open" in moduleData.Connection.prototype) {
        var oldOpen = moduleData.Connection.prototype.open;

        moduleData.Connection.prototype.open = function () {
          var args = [];
          args.push(props.url);
          var originalOptions = null;
          var callback = null;
          // Arguments can be in different length, url, options, callback
          if (typeof arguments[arguments.length - 1] === "function") {
            callback = arguments[arguments.length - 1];
            if (typeof arguments[arguments.length - 2] === "object") {
              originalOptions = arguments[arguments.length - 2];
            }
          }
          else if (typeof arguments[arguments.length - 1] === "object") {
            originalOptions = arguments[arguments.length - 1];
          }

          if (originalOptions !== null) {
            args.push(originalOptions);
          }

          if (callback !== null) {
            args.push(callback);
          }

          return oldOpen.apply(this, args);
        };
      }
    }
  };
}

util.inherits(Mongodb, sc.Service);

function setupMongodb () {
  var service = new Mongodb();
  service.setup();
}

exports.setup = setupMongodb;
