// cf-autoconfig/lib/mysql.js
/*
 * Cloud Foundry auto-configuration
 * Mysql connection
 *
 * Auto-configured node modules:
 *
 * "mysql" [https://github.com/felixge/node-mysql]
 * - createClient()
 * - Client._connect()
 */

var sc = require(require("path").join(__dirname, "service"));
var util = require("util");

function Mysql () {
  this.type = "mysql";
  this.supportedModules = {
    "mysql" : function (moduleData, props) {

      // createClient()
      if ("createClient" in moduleData) {
        var oldCreateClient = moduleData.createClient;
        var oldCreateClientProto = moduleData.createClient.prototype;

        moduleData.createClient = function () {
          var options = arguments[0];
          options.host = props.host;
          options.port = props.port;
          options.user = props.username;
          options.password = props.password;
          options.database = props.database;
          return oldCreateClient.call(this, options);
        };
        moduleData.createClient.prototype = oldCreateClientProto;
      }

      // Client._connect()
      if ("Client" in moduleData && "_connect" in moduleData.Client.prototype) {
        var oldClientConnect = moduleData.Client.prototype._connect;
        moduleData.Client.prototype._connect = function () {
          this.host = props.host;
          this.port = props.port;
          this.user = props.username;
          this.password = props.password;
          this.database = props.database;
          return oldClientConnect.call(this);
        };
      }
    }
  };
}

util.inherits(Mysql, sc.Service);

function setupMysql () {
  var service = new Mysql();
  service.setup();
}

exports.setup = setupMysql;
