// cf-autoconfig/lib/postgresql.js
/*
 * Cloud Foundry auto-configuration
 * Postgresql connection
 *
 * Auto-configured node modules:
 *
 * "pg" [https://github.com/brianc/node-postgres]
 * - defaults
 * - connect()
 * - new Client()
 */

var sc = require(require("path").join(__dirname, "service"));
var util = require("util");

function Postgresql () {
  this.type = "postgresql";
  this.supportedModules = {
    "pg" : function (moduleData, props) {
      // defaults
      if (moduleData.hasOwnProperty("defaults")) {
        moduleData.defaults = processConfig(moduleData.defaults, props);
      }

      // connect()
      // this module returns an object, not function, using getPrototypeOf
      if (typeof moduleData === "object") {
        var pgProto = Object.getPrototypeOf(moduleData);
        if ("connect" in pgProto) {
          var oldConnect = pgProto.connect;
          pgProto.connect = function () {
            var args = Array.prototype.slice.call(arguments);
            args[0] = processConfig(args[0], props);
            return oldConnect.apply(this, args);
          };
        }
      }

      // new Client()
      if ("Client" in moduleData) {
        var oldClient = moduleData.Client;
        var oldClientProto = moduleData.Client.prototype;

        moduleData.Client = function () {
          var self = function () {
            var args = Array.prototype.slice.call(arguments);
            args[0] = processConfig(args[0], props);
            return oldClient.apply(this, args);
          };

          for (var prop in oldClient) {
            self[prop] = oldClient[prop];
          }

          return self;
        }();
        moduleData.Client.prototype = oldClientProto;
      }
    }
  };
}

util.inherits(Postgresql, sc.Service);

function setupPostgresql () {
  var service = new Postgresql();
  service.setup();
}

exports.setup = setupPostgresql;

function processConfig(config, props) {
  if (typeof config === "string") {
    config = props.url;
  }
  else {
    config = config || {};
    config.host = props.host;
    config.port = props.port;
    config.user = props.username;
    config.password = props.password;
    config.database = props.database;
  }
  return config;
}
