// cf-autoconfig/lib/redis.js
/*
 * Cloud Foundry auto-configuration
 * Redis connection
 *
 * Auto-configured node modules:
 *
 * "redis" [https://github.com/mranney/node_redis]
 * - createClient()
 * - auth()
 */

var path = require("path");
var sc = require(path.resolve(path.join(__dirname, "service")));
var util = require("util");

function Redis () {
  this.type = "redis";
  this.supportedModules = {
    "redis" : function (moduleData, props) {
      // auth()
      if ("RedisClient" in moduleData && "auth" in moduleData.RedisClient.prototype) {
        var oldAuth = moduleData.RedisClient.prototype.auth;

        moduleData.RedisClient.prototype.auth = function () {
          var args = Array.prototype.slice.call(arguments);
          args[0] = props.password;
          return oldAuth.apply(this, args);
        };
      }

      // createClient()
      if ("createClient" in moduleData) {
        var oldConnect = moduleData.createClient;
        var oldConnectProto = moduleData.createClient.prototype;

        moduleData.createClient = function () {
          var args = Array.prototype.slice.call(arguments);
          args[0] = props.port;
          args[1] = props.host;
          var client = oldConnect.apply(this, args);
          if (props.password !== null) client.auth_pass = props.password;
          return client;
        };

        moduleData.createClient.prototype = oldConnectProto;
      }
    }
  };
}

util.inherits(Redis, sc.Service);

function setupRedis () {
  var service = new Redis();
  service.setup();
}

exports.setup = setupRedis;
